### PR TITLE
ARROW-5150: [Ruby] Add Arrow::Table#raw_records

### PR DESCRIPTION
**Pull request by mrkn**
I'd like to add `raw_records` method in `Arrow::Table` class.
The same method of `RecordBatch` was added in #3587.

### CODE REVIEW

File: ruby/red-arrow/lib/arrow/data-type.rb [hunk 1]
```diff
@@ -18,6 +18,17 @@
 module Arrow
   class DataType
     class << self
+      # Return the corresponding array class
+      #
+      # @return [Class, nil] The corresponding array class.
+      #    If there is no corresponding array class, it returns nil.
+      def array_type
```
Comment (mrkn) [inline ruby/red-arrow/lib/arrow/data-type.rb]: @kou How do you think about this new API?

Comment (mrkn) [reply ruby/red-arrow/lib/arrow/data-type.rb]: Ah, Sorry.  I should remove this because I've never use this in test code...

**Comment by mrkn** [comment]
@kou I think this PR is ready to review.  Please have a look.

File: ruby/red-arrow/ext/arrow/raw-records.cpp [hunk 2]
```diff
@@ -642,11 +642,35 @@ namespace red_arrow {
...
+          const auto n_rows = table.num_rows();
+          for (int64_t i = 0; i < n_rows; ++i) {
+            auto record = rb_ary_new_capa(n_columns_);
+            rb_ary_push(records_, record);
+          }
+          for (int i = 0; i < n_columns_; ++i) {
+            const auto column = table.column(i).get();
+            const auto chunked_array = column->data();
+            column_index_ = i;
+            // TODO: should we use array->offset() ?
```
Comment (kou) [inline ruby/red-arrow/ext/arrow/raw-records.cpp]: `array->offset()` is for sliced array. So we can use it for this case.

File: ruby/red-arrow/ext/arrow/raw-records.cpp [hunk 3]
```diff
@@ -642,11 +642,35 @@ namespace red_arrow {
             auto record = rb_ary_new_capa(n_columns_);
             rb_ary_push(records_, record);
           }
+          row_offset_ = 0;
           for (int i = 0; i < n_columns_; ++i) {
             const auto array = record_batch.column(i).get();
             column_index_ = i;
             check_status(array->Accept(this),
-                         "[raw-records]");
+                         "[record-batch-raw-records]");
```
Comment (kou) [inline ruby/red-arrow/ext/arrow/raw-records.cpp]: Could you use `[record-batch][raw-records]`?
We use multiple tags to show nested contexts.

**Review by kou — CHANGES_REQUESTED** [review]
Could you check my comments?

We can share tests by including test module in test-unit.
I'll push the change for it. Could you fix others?

**Comment by mrkn** [comment]
> I'll push the change for it. Could you fix others?

Yes!

**Review by kou — APPROVED** [review]
+1
I'll merge this when CI is passed.

**Comment by kou** [comment]
Java job failure is unrelated.

```text
$ ~/bin/install-jdk.sh --target "/home/travis/openjdk9" --workspace "/home/travis/.cache/install-jdk" --feature "9" --license "GPL" --cacerts
install-jdk.sh 2019-04-18
The command "~/bin/install-jdk.sh --target "/home/travis/openjdk9" --workspace "/home/travis/.cache/install-jdk" --feature "9" --license "GPL" --cacerts" failed and exited with 51 during .
```

**Pull request closed without merging**